modules/build/src/main/scala/scala/build/internal/markdown/MarkdownCodeWrapper.scala

```diff
@@ -0,0 +1,156 @@
+package scala.build.internal.markdown
+
+import scala.annotation.tailrec
+import scala.build.internal.AmmUtil
+import scala.build.internal.markdown.MarkdownCodeBlockUtil.MarkdownFence
+import scala.build.internal.Name
+
+/** A util for extraction and wrapping of code blocks in Markdown files.
+  */
+object MarkdownCodeWrapper {
+
+  /** Extracts scala code blocks from Markdown snippets, divides them into 3 categories and wraps
+    * when necessary.
+    *
+    * @param subPath
+    *   the project [[os.SubPath]] to the Markdown file
+    * @param content
+    *   Markdown code
+    * @return
+    *   a tuple of (Option(simple scala snippets code), Option(raw scala snippets code), Option(test
+    *   scala snippets code))
+    */
+  def apply(
+    subPath: os.SubPath,
+    content: String
+  ): (Option[String], Option[String], Option[String]) = {
+    val (pkg, wrapper) = AmmUtil.pathToPackageWrapper(subPath)
+    val maybePkgString =
+      if pkg.isEmpty then None else Some(s"package ${AmmUtil.encodeScalaSourcePath(pkg)}")
+    val allSnippets                      = MarkdownCodeBlockUtil.findFences(content)
+    val (rawSnippets, processedSnippets) = allSnippets.partition(_.isRaw)
+    val (testSnippets, mainSnippets)     = processedSnippets.partition(_.isTest)
+    val wrapperName                      = s"${wrapper.raw}_md"
+    (
+      wrapScalaCode(mainSnippets, wrapperName, maybePkgString),
+      rawScalaCode(rawSnippets),
+      rawScalaCode(testSnippets)
+    )
+  }
+
+  /** Scope object name for a given index
```

Comment:
Can you explain more, how works index? Is this a scope ID that can be reset by `reset` keyword?

Reply:
yep, it's the scope id

Reply:
BTW we might have to get rid of the scopes completely when fixing the using directives issues in a later PR

Reply:
Added some more scaladocs, should be clearer now